File: glisse.py
import sys
import functools
import json
import os
import subprocess
import tomllib
from dataclasses import asdict, dataclass, field
from typing import Callable, List, Optional

# --- exceptions ---


class GlisseConfigError(BaseException):
    pass


class GlisseConfigLoadingError(BaseException):
    pass


# --- eDSL Core ---


class Branch:
    def __init__(self, name: str):
        self.name = name
        self.hooks: List[Callable] = []
        self.next_branch: Optional[Branch] = None

    def when_merged(self, func: Callable):
        self.hooks.append(func)
        return self

    def __gt__(self, other: "Branch"):
        """Overloads the '>' operator to define order."""
        self.next_branch = other
        return other

    def __repr__(self):
        return f"Branch({self.name})"


# --- State Persistence (Same Logic as Phase 1) ---


@dataclass
class MergeStep:
    target_branch: str
    original_sha: str
    tags_created: List[str] = field(default_factory=list)


STATE_FILE = ".merge_state.json"


class DSLRunner:
    STATE_FILE = STATE_FILE

    def __init__(self, start_node: Branch):
        self.start_node = start_node
        self.history: List[MergeStep] = []

    def _get_pipeline(self) -> List[Branch]:
        nodes = []
        curr = self.start_node
        while curr:
            nodes.append(curr)
            curr = curr.next_branch
        return nodes

    def _save_state(self):
        with open(self.STATE_FILE, "w") as f:
            json.dump([asdict(s) for s in self.history], f, indent=2)

    def execute(self):
        pipeline = self._get_pipeline()

        for i in range(len(pipeline) - 1):
            src, tgt = pipeline[i], pipeline[i + 1]
            print(f"\n>>> Merging {src.name} -> {tgt.name}")

            # Capture state
            pre_tags = self._get_tags()
            target_sha = self._get_sha(tgt.name)

            step = MergeStep(target_branch=tgt.name, original_sha=target_sha)
            self.history.append(step)
            self._save_state()

            try:
                # Execution
                self._git(["checkout", tgt.name])
                self._git(["merge", src.name, "--no-ff", "-m", f"Merge {src.name}"])

                # Update tags in state
                step.tags_created = list(self._get_tags() - pre_tags)
                self._save_state()

                # Run eDSL hooks
                for hook in tgt.hooks:
                    hook({"step": step, "source": src.name})

            except Exception as e:
                print(f"Error during merge to {tgt.name}: {e}")
                sys.exit(1)

        print("\nPipeline Complete.")
        # if os.path.exists(self.STATE_FILE): os.remove(self.STATE_FILE)

    def unwind(self):
        if not os.path.exists(self.STATE_FILE):
            print("Nothing to unwind.")
            return

        with open(self.STATE_FILE, "r") as f:
            data = json.load(f)
            steps = [MergeStep(**d) for d in data]

        for step in reversed(steps):
            print(f"Rolling back {step.target_branch}...")
            for tag in step.tags_created:
                self._git(["tag", "-d", tag])
            self._git(["checkout", step.target_branch])
            self._git(["reset", "--hard", step.original_sha])

        os.remove(self.STATE_FILE)
        print("Unwind complete.")

    # Helpers
    def _git(self, cmd):
        return subprocess.run(["git"] + cmd, check=True, capture_output=True, text=True)

    def _get_sha(self, b):
        return self._git(["rev-parse", b]).stdout.strip()

    def _get_tags(self):
        return set(self._git(["tag"]).stdout.splitlines())


# --- User Script ---


def test():
    dev = Branch("dev")
    staging = Branch("staging")
    main = Branch("main")

    # The eDSL definition
    dev > staging > main

    staging.when_merged(lambda ctx: os.system("echo 'Bump staging version'"))
    main.when_merged(lambda ctx: os.system("echo 'Bump main version'"))
    return [dev, staging, main]


def load_from_config():
    try:
        with open("./pyproject.toml", "rb") as fh:
            pyproject = tomllib.load(fh)
        config = pyproject["glisse"]
        order = config["order"]
        branches = [Branch(b) for b in order]
        functools.reduce(lambda b, c: b > c, branches)
        for branch in branches:
            try:
                for hook in config[branch.name]["merged"]["hooks"]:
                    branch.when_merged(lambda ctx: os.system(hook))
            except KeyError:
                # no merge hook for branch
                pass
        return branches
    except KeyError as e:
        raise GlisseConfigLoadingError(f"config missing key {e}")
    except FileNotFoundError as e:
        raise GlisseConfigLoadingError(f"file not found {e}")


def show_config():
    branches = load_from_config()
    print("\t current branch: TODO")
    print(f"\t order: {' > '.join(b.name for b in branches)}")
    print("\t hooks: TODO")
    if os.path.exists(STATE_FILE):
        with open(STATE_FILE, "r") as f:
            data = json.load(f)
            steps = [MergeStep(**d) for d in data]
            print("\t state:")
            for step in steps:
                print(f"\t\t {step.original_sha} {step.target_branch}")

File: glisse.rs
use std::collections::HashSet;
use std::fs::{File, remove_file};
use std::io::{Write, BufReader};
use std::process::{Command, exit};
use serde::{Serialize, Deserialize};
use serde_json;

// --- eDSL Core ---

type Hook = Box<dyn Fn(&MergeContext)>;

#[derive(Clone)]
struct Branch {
    name: String,
    hooks: Vec<Hook>,
    next_branch: Option<Box<Branch>>,
}

impl Branch {
    fn new(name: &str) -> Self {
        Branch {
            name: name.to_string(),
            hooks: Vec::new(),
            next_branch: None,
        }
    }

    fn when_merged<F>(&mut self, func: F) -> &mut Self
    where
        F: Fn(&MergeContext) + 'static,
    {
        self.hooks.push(Box::new(func));
        self
    }

    fn then(mut self, next: Branch) -> Branch {
        self.next_branch = Some(Box::new(next));
        *self.next_branch.as_mut().unwrap()
    }
}

// --- State Persistence ---

#[derive(Serialize, Deserialize)]
struct MergeStep {
    target_branch: String,
    original_sha: String,
    tags_created: Vec<String>,
}

struct MergeContext<'a> {
    step: &'a MergeStep,
    source: &'a str,
}

struct DSLRunner {
    start_node: Branch,
    history: Vec<MergeStep>,
}

impl DSLRunner {
    const STATE_FILE: &'static str = ".merge_state.json";

    fn new(start_node: Branch) -> Self {
        DSLRunner {
            start_node,
            history: Vec::new(),
        }
    }

    fn get_pipeline(&self) -> Vec<&Branch> {
        let mut pipeline = Vec::new();
        let mut current = Some(&self.start_node);

        while let Some(branch) = current {
            pipeline.push(branch);
            current = branch.next_branch.as_deref();
        }

        pipeline
    }

    fn save_state(&self) {
        let json = serde_json::to_string_pretty(&self.history).unwrap();
        let mut file = File::create(Self::STATE_FILE).unwrap();
        file.write_all(json.as_bytes()).unwrap();
    }

    fn execute(&mut self) {
        let pipeline = self.get_pipeline();

        for i in 0..pipeline.len() - 1 {
            let src = pipeline[i];
            let tgt = pipeline[i+1];

            println!("\n>>> Merging {} -> {}", src.name, tgt.name);

            let pre_tags = self.get_tags();
            let target_sha = self.get_sha(&tgt.name);

            let mut step = MergeStep {
                target_branch: tgt.name.clone(),
                original_sha: target_sha,
                tags_created: Vec::new(),
            };

            self.history.push(step.clone());
            self.save_state();

            // Merge
            if let Err(e) = self.git(&["checkout", &tgt.name]) {
                eprintln!("Error: {}", e);
                exit(1);
            }
            if let Err(e) = self.git(&["merge", &src.name, "--no-ff", "-m", &format!("Merge {}", src.name)]) {
                eprintln!("Error: {}", e);
                exit(1);
            }

            let post_tags = self.get_tags();
            step.tags_created = post_tags.difference(&pre_tags).cloned().collect();
            self.history.last_mut().unwrap().tags_created = step.tags_created.clone();
            self.save_state();

            // Run hooks
            for hook in &tgt.hooks {
                let ctx = MergeContext { step: &step, source: &src.name };
                hook(&ctx);
            }
        }

        println!("\nPipeline Complete.");
    }

    fn unwind(&self) {
        if !std::path::Path::new(Self::STATE_FILE).exists() {
            println!("Nothing to unwind.");
            return;
        }

        let file = File::open(Self::STATE_FILE).unwrap();
        let reader = BufReader::new(file);
        let steps: Vec<MergeStep> = serde_json::from_reader(reader).unwrap();

        for step in steps.iter().rev() {
            println!("Rolling back {}...", step.target_branch);
            for tag in &step.tags_created {
                let _ = self.git(&["tag", "-d", tag]);
            }
            let _ = self.git(&["checkout", &step.target_branch]);
            let _ = self.git(&["reset", "--hard", &step.original_sha]);
        }

        let _ = remove_file(Self::STATE_FILE);
        println!("Unwind complete.");
    }

    fn git(&self, args: &[&str]) -> Result<(), String> {
        let output = Command::new("git")
            .args(args)
            .output()
            .map_err(|e| e.to_string())?;

        if !output.status.success() {
            Err(String::from_utf8_lossy(&output.stderr).to_string())
        } else {
            Ok(())
        }
    }

    fn get_sha(&self, branch: &str) -> String {
        let output = Command::new("git")
            .args(&["rev-parse", branch])
            .output()
            .unwrap();
        String::from_utf8_lossy(&output.stdout).trim().to_string()
    }

    fn get_tags(&self) -> HashSet<String> {
        let output = Command::new("git")
            .args(&["tag"])
            .output()
            .unwrap();
        output.stdout.lines()
            .map(|l| l.unwrap().to_string())
            .collect()
    }
}

// --- User Script ---

fn main() {
    let mut dev = Branch::new("dev");
    let mut staging = Branch::new("staging");
    let mut main = Branch::new("main");

    dev.next_branch = Some(Box::new(staging.clone()));
    staging.next_branch = Some(Box::new(main.clone()));

    staging.when_merged(|_ctx| {
        let _ = Command::new("sh")
            .arg("-c")
            .arg("echo 'Bump staging version'")
            .status();
    });

    main.when_merged(|_ctx| {
        let _ = Command::new("sh")
            .arg("-c")
            .arg("echo 'Bump main version'")
            .status();
    });

    let mut runner = DSLRunner::new(dev);

    let args: Vec<String> = std::env::args().collect();
    if args.contains(&"--undo".to_string()) {
        runner.unwind();
    } else {
        runner.execute();
    }
}
